aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -892,6 +920,9 @@ void take_out_cpu_template(
     auto index_continuous = index.contiguous();
     bool is_contiguous = input.is_contiguous();
     auto input_size = input.numel();
+    at::assert_no_internal_overlap(output);
+    at::assert_no_partial_overlap(output, index);
```

Comment:
For normal element-wise kernels, we use `assert_no_partial_overlap` because everything is read at a given index, then the corresponding index output is written after. 
```python
for i in np.ndindex(...):
    output[i] = f(input[i])
```

Full overlap is okay because we never read a given index after writing to it. For indexing operations, the output and input are written/read at different indexes, so full overlap isn't okay.

For `take`, case we're basically doing:
```python
for i in np.ndindex(...):
    output[i] = input[index[i]]
```
`output` and `index` are always accessed at the same index so full overlap is okay. Whereas the `input` is read in an unknown order depending on the value of `index`, so full overlap isn't okay.

Reply:
By that logic, `gather` should also only assert `no_partial_overlap` between output and index? `gather` and `take` are conceptually very similar operations.

Reply:
Good point, I've changed `gather`.